File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = word
  end

  def match(words)
    sorted_chars = @word.downcase.chars.sort

    words.select do |candidate|
      candidate.downcase != @word.downcase &&
        candidate.downcase.chars.sort == sorted_chars
    end
  end
end

File: go/anagram/anagram.go
package anagram

import (
	"sort"
	"strings"
)

// Detect returns a subset of candidates that are anagrams of subject.
func Detect(subject string, candidates []string) (result []string) {
	sortedChars := func(input string) string {
		input = strings.ToLower(input)
		chars := strings.Split(input, "")
		sort.Strings(chars)
		return strings.Join(chars, "")
	}

	sortedSubject := sortedChars(subject)

	for _, candidate := range candidates {
		candidate = strings.ToLower(candidate)
		sorted := sortedChars(candidate)

		if sortedSubject == sorted && subject != candidate {
			result = append(result, candidate)
		}
	}

	return
}
